api/src/pcapi/alembic/versions/20241220T095131_4c3be4ff5274_add_timespan_on_headline_offer.py

```diff
@@ -0,0 +1,59 @@
+"""
+add timespan on headline_offer table
+"""
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+
+# pre/post deployment: post
+# revision identifiers, used by Alembic.
+revision = "4c3be4ff5274"
+down_revision = "4c53718279e7"
+branch_labels: tuple[str] | None = None
+depends_on: list[str] | None = None
+
+
+def upgrade() -> None:
+    op.add_column("headline_offer", sa.Column("timespan", postgresql.TSRANGE(), nullable=False))
+    op.execute(
+        """
+        ALTER TABLE headline_offer
+            ADD CONSTRAINT exclude_offer_timespan EXCLUDE USING gist (
+                "offerId"
+                WITH
+                    =,
+                    timespan
+                WITH
+                    &&
+            )
```

Comment:
Le Add Constraint prends un lock un peu énervé et sera bloqué par les select sur cette table. Si on a une requête qui fait une jointure sur cette table (et je vois qu'il en éxiste déjà), on aura surement un problème pour prendre le lock. Il faut les créer en NOT VALID et dans une migration suivante valider la contrainte

Reply:
Malheureusement, une contrainte EXCLUDE ne peut pas être créé en NOT VALID, ce n'est pas supporté par postgresql

Reply:
```
sqlalchemy.exc.NotSupportedError: (psycopg2.errors.FeatureNotSupported) EXCLUDE constraints cannot be marked NOT VALID

[SQL: ALTER TABLE headline_offer ADD CONSTRAINT exclude_offer_timespan EXCLUDE USING gist ("offerId" WITH =, timespan WITH &&) NOT VALID]
```